File: pkg/store/conversation_store.go
package store

import (
	"context"
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// ConversationStore provides data access operations for the Conversation model.
type ConversationStore struct {
	db *gorm.DB
}

// Create inserts a new conversation record into the database.
func (cs *ConversationStore) Create(ctx context.Context, conv *model.Conversation) error {
	if err := cs.db.WithContext(ctx).Create(conv).Error; err != nil {
		return classifyError(fmt.Errorf("store: create conversation: %w", err))
	}
	return nil
}

// Get retrieves a conversation by its primary key. Returns ErrNotFound if no
// record exists.
func (cs *ConversationStore) Get(ctx context.Context, id string) (*model.Conversation, error) {
	var conv model.Conversation
	err := cs.db.WithContext(ctx).
		Where("id = ?", id).
		First(&conv).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ErrNotFound
		}
		return nil, classifyError(fmt.Errorf("store: get conversation: %w", err))
	}
	return &conv, nil
}

// GetByUsers returns the 1-on-1 conversation between user1 and user2.
// It checks both (user1, user2) and (user2, user1) orderings.
// Returns ErrNotFound if no matching conversation exists.
func (cs *ConversationStore) GetByUsers(ctx context.Context, user1, user2 string) (*model.Conversation, error) {
	var conv model.Conversation
	err := cs.db.WithContext(ctx).
		Where("(user_id1 = ? AND user_id2 = ?) OR (user_id1 = ? AND user_id2 = ?)", user1, user2, user2, user1).
		First(&conv).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ErrNotFound
		}
		return nil, classifyError(fmt.Errorf("store: get conversation by users: %w", err))
	}
	return &conv, nil
}

// GetByUser returns conversations where the given user is either UserID1 or
// UserID2, ordered by LastMessageAt descending, with offset/limit pagination.
// Soft-deleted records are excluded automatically by GORM's soft-delete plugin.
func (cs *ConversationStore) GetByUser(ctx context.Context, userID string, offset, limit int) ([]*model.Conversation, error) {
	if limit <= 0 || limit > 101 {
		limit = 20
	}
	if offset < 0 {
		offset = 0
	}

	var convs []*model.Conversation
	if err := cs.db.WithContext(ctx).
		Where("(user_id1 = ? OR user_id2 = ?) AND user_id2 != ''", userID, userID).
		Order("last_message_at DESC").
		Offset(offset).
		Limit(limit).
		Find(&convs).Error; err != nil {
		return nil, classifyError(fmt.Errorf("store: get conversations by user: %w", err))
	}
	return convs, nil
}

// Update saves all fields of the conversation back to the database.
// It uses Unscoped() so that soft-deleted records can be updated (including
// clearing deleted_at to NULL when restoring a conversation).
func (cs *ConversationStore) Update(ctx context.Context, conv *model.Conversation) error {
	if err := cs.db.WithContext(ctx).Unscoped().Save(conv).Error; err != nil {
		return classifyError(err)
	}
	return nil
}

// Upsert creates the conversation if it does not exist, or saves (overwrites)
// it if it already exists. This is used by the client sync pipeline to apply
// conversation create events idempotently (D-045).
// It uses Unscoped() to also find soft-deleted records, so that restoring a
// previously deleted conversation correctly transitions it back to active.
// If a concurrent insert causes a duplicate key error, the operation retries
// as an update to handle the TOCTOU race between SELECT and INSERT.
func (cs *ConversationStore) Upsert(ctx context.Context, conv *model.Conversation) error {
	var existing model.Conversation
	err := cs.db.WithContext(ctx).Unscoped().Where("id = ?", conv.ID).First(&existing).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			if createErr := cs.db.WithContext(ctx).Create(conv).Error; createErr != nil {
				// TOCTOU: another goroutine may have inserted between SELECT and INSERT.
				if errors.Is(classifyError(createErr), ErrDuplicateKey) {
					return cs.Update(ctx, conv)
				}
				return classifyError(fmt.Errorf("store: upsert conversation: %w", createErr))
			}
			return nil
		}
		return classifyError(fmt.Errorf("store: upsert conversation: %w", err))
	}
	return cs.Update(ctx, conv)
}

// Delete performs a cascading soft delete: the conversation and all its messages
// are soft-deleted within a single transaction (D-013).
func (cs *ConversationStore) Delete(ctx context.Context, id string) error {
	return cs.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		result := tx.Delete(&model.Conversation{}, "id = ?", id)
		if result.Error != nil {
			return classifyError(fmt.Errorf("store: delete conversation: %w", result.Error))
		}
		if result.RowsAffected == 0 {
			return ErrNotFound
		}

		// Cascade soft-delete all messages in this conversation (D-013).
		if err := tx.Where("conversation_id = ?", id).Delete(&model.Message{}).Error; err != nil {
			return classifyError(fmt.Errorf("store: cascade delete messages: %w", err))
		}
		return nil
	})
}

// Restore undeletes a soft-deleted conversation and cascades the restore to all
// its messages within a single transaction (D-015). Calling Restore on a
// conversation that already exists but is not soft-deleted is idempotent — it
// returns nil without error (D-015). Returns ErrNotFound only if the
// conversation does not exist at all.
func (cs *ConversationStore) Restore(ctx context.Context, id string) error {
	return cs.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		// Check if conversation exists at all (including soft-deleted).
		var count int64
		if err := tx.Unscoped().Model(&model.Conversation{}).Where("id = ?", id).Count(&count).Error; err != nil {
			return classifyError(fmt.Errorf("store: restore conversation: %w", err))
		}
		if count == 0 {
			return ErrNotFound
		}

		// Restore the conversation if it was soft-deleted.
		result := tx.Unscoped().
			Model(&model.Conversation{}).
			Where("id = ? AND deleted_at IS NOT NULL", id).
			Update("deleted_at", nil)
		if result.Error != nil {
			return classifyError(fmt.Errorf("store: restore conversation: %w", result.Error))
		}

		// Cascade restore all messages in this conversation (D-015).
		// Only runs if the conversation was actually soft-deleted (RowsAffected > 0),
		// making this idempotent for non-deleted conversations.
		if result.RowsAffected > 0 {
			if err := tx.Unscoped().
				Model(&model.Message{}).
				Where("conversation_id = ? AND deleted_at IS NOT NULL", id).
				Update("deleted_at", nil).Error; err != nil {
				return classifyError(fmt.Errorf("store: cascade restore messages: %w", err))
			}
		}
		return nil
	})
}

// UpdateLastMessage updates the LastMessageAt and LastProcessedMessageID fields
// of the conversation identified by convID.
func (cs *ConversationStore) UpdateLastMessage(ctx context.Context, convID string, lastMessageAt time.Time, lastProcessedMessageID uint32) error {
	result := cs.db.WithContext(ctx).
		Model(&model.Conversation{}).
		Where("id = ?", convID).
		Updates(map[string]any{
			"last_message_at":           lastMessageAt,
			"last_processed_message_id": lastProcessedMessageID,
		})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: update last message: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// UpdateLastRead updates the last-read message ID for the specified user.
// Uses MAX semantics: only advances forward, never backward (D-012).
// Uses a single SQL statement to avoid TOCTOU races.
func (cs *ConversationStore) UpdateLastRead(ctx context.Context, convID, userID string, messageID uint32) error {
	// Single UPDATE with CASE WHEN for both columns; only the matching user column advances.
	// WHERE clause ensures the user belongs to this conversation and RowsAffected catches missing records.
	result := cs.db.WithContext(ctx).
		Model(&model.Conversation{}).
		Where("id = ? AND (user_id1 = ? OR user_id2 = ?)", convID, userID, userID).
		Updates(map[string]any{
			"last_read_message_id1": gorm.Expr("CASE WHEN user_id1 = ? AND ? > last_read_message_id1 THEN ? ELSE last_read_message_id1 END", userID, messageID, messageID),
			"last_read_message_id2": gorm.Expr("CASE WHEN user_id2 = ? AND ? > last_read_message_id2 THEN ? ELSE last_read_message_id2 END", userID, messageID, messageID),
		})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: update last read: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// SearchByTitle searches conversations for the given user that contain the
// specified title substring (case-insensitive via LIKE), ordered by
// LastMessageAt descending.
func (cs *ConversationStore) SearchByTitle(ctx context.Context, userID, title string, limit int) ([]*model.Conversation, error) {
	if limit <= 0 || limit > 101 {
		limit = 20
	}
	if title == "" {
		return []*model.Conversation{}, nil
	}

	like := "%" + escapeLikePattern(title) + "%"

	var convs []*model.Conversation
	err := cs.db.WithContext(ctx).
		Where("(user_id1 = ? OR user_id2 = ?) AND title LIKE ? ESCAPE '|'", userID, userID, like).
		Order("last_message_at DESC").
		Limit(limit).
		Find(&convs).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: search conversations by title: %w", err))
	}
	return convs, nil
}

// GetUnscoped retrieves a conversation including soft-deleted records.
// Returns ErrNotFound if no record exists.
func (cs *ConversationStore) GetUnscoped(ctx context.Context, id string) (*model.Conversation, error) {
	var conv model.Conversation
	err := cs.db.WithContext(ctx).
		Unscoped().
		Where("id = ?", id).
		First(&conv).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ErrNotFound
		}
		return nil, classifyError(fmt.Errorf("store: get unscoped conversation: %w", err))
	}
	return &conv, nil
}

// UpdateLastMessageTx updates last message fields within the given transaction.
func (cs *ConversationStore) UpdateLastMessageTx(ctx context.Context, tx *gorm.DB, convID string, lastMessageAt time.Time, lastProcessedMessageID uint32) error {
	result := tx.WithContext(ctx).
		Model(&model.Conversation{}).
		Where("id = ?", convID).
		Updates(map[string]any{
			"last_message_at":           lastMessageAt,
			"last_processed_message_id": lastProcessedMessageID,
		})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: update last message tx: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// UpdateLastReadTx updates read cursor within the given transaction.
// Uses MAX semantics: only advances forward (D-012).
func (cs *ConversationStore) UpdateLastReadTx(ctx context.Context, tx *gorm.DB, convID, userID string, messageID uint32) error {
	result := tx.WithContext(ctx).
		Model(&model.Conversation{}).
		Where("id = ? AND (user_id1 = ? OR user_id2 = ?)", convID, userID, userID).
		Updates(map[string]any{
			"last_read_message_id1": gorm.Expr("CASE WHEN user_id1 = ? AND ? > last_read_message_id1 THEN ? ELSE last_read_message_id1 END", userID, messageID, messageID),
			"last_read_message_id2": gorm.Expr("CASE WHEN user_id2 = ? AND ? > last_read_message_id2 THEN ? ELSE last_read_message_id2 END", userID, messageID, messageID),
		})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: update last read tx: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// UpsertTx creates or updates a conversation within the given transaction.
// It uses Unscoped() to also find soft-deleted records, so that restoring a
// previously deleted conversation correctly transitions it back to active.
// If a concurrent insert causes a duplicate key error, the operation retries
// as an update to handle the TOCTOU race between SELECT and INSERT.
func (cs *ConversationStore) UpsertTx(ctx context.Context, tx *gorm.DB, conv *model.Conversation) error {
	var existing model.Conversation
	err := tx.WithContext(ctx).Unscoped().Where("id = ?", conv.ID).First(&existing).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			if createErr := tx.WithContext(ctx).Create(conv).Error; createErr != nil {
				// TOCTOU: another goroutine may have inserted between SELECT and INSERT.
				// Use classifyError to detect duplicate key via SQLite error strings.
				if errors.Is(classifyError(createErr), ErrDuplicateKey) {
					return cs.updateExistingTx(ctx, tx, conv)
				}
				return classifyError(fmt.Errorf("store: upsert create tx: %w", createErr))
			}
			return nil
		}
		return classifyError(fmt.Errorf("store: upsert tx: %w", err))
	}
	return cs.updateExistingTx(ctx, tx, conv)
}

// updateExistingTx saves the conversation record, preserving server-controlled fields.
// It uses Unscoped() so that soft-deleted records can be updated (including
// clearing deleted_at to NULL when restoring a conversation).
func (cs *ConversationStore) updateExistingTx(ctx context.Context, tx *gorm.DB, conv *model.Conversation) error {
	if err := tx.WithContext(ctx).Unscoped().Save(conv).Error; err != nil {
		return classifyError(fmt.Errorf("store: upsert update tx: %w", err))
	}
	return nil
}

// SoftDeleteTx performs cascading soft delete within the given transaction (D-013).
func (cs *ConversationStore) SoftDeleteTx(ctx context.Context, tx *gorm.DB, id string) error {
	result := tx.WithContext(ctx).Delete(&model.Conversation{}, "id = ?", id)
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: delete conversation tx: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	// Cascade soft-delete messages (D-013).
	if err := tx.WithContext(ctx).Where("conversation_id = ?", id).Delete(&model.Message{}).Error; err != nil {
		return classifyError(fmt.Errorf("store: cascade delete messages tx: %w", err))
	}
	return nil
}

// RestoreTx performs cascading restore within the given transaction (D-015).
func (cs *ConversationStore) RestoreTx(ctx context.Context, tx *gorm.DB, id string) error {
	// Check existence (including soft-deleted).
	var count int64
	if err := tx.WithContext(ctx).Unscoped().Model(&model.Conversation{}).Where("id = ?", id).Count(&count).Error; err != nil {
		return classifyError(fmt.Errorf("store: restore check tx: %w", err))
	}
	if count == 0 {
		return ErrNotFound
	}
	// Restore conversation if soft-deleted.
	result := tx.WithContext(ctx).Unscoped().
		Model(&model.Conversation{}).
		Where("id = ? AND deleted_at IS NOT NULL", id).
		Update("deleted_at", nil)
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: restore conversation tx: %w", result.Error))
	}
	// Cascade restore messages (D-015).
	if result.RowsAffected > 0 {
		if err := tx.WithContext(ctx).Unscoped().
			Model(&model.Message{}).
			Where("conversation_id = ? AND deleted_at IS NOT NULL", id).
			Update("deleted_at", nil).Error; err != nil {
			return classifyError(fmt.Errorf("store: cascade restore messages tx: %w", err))
		}
	}
	return nil
}

File: demo/vue-pure-admin/packages/xyncra-client-core/src/db/conversation-store.ts
/**
 * ConversationStore — data access operations for the Conversation model.
 *
 * Mirrors Go ConversationStore (pkg/store/conversation_store.go).
 *
 * Key constraints:
 *   C4 — UpdateLastRead uses MAX semantics (only advances forward, never backward).
 *   Soft delete — deleted_at field is used instead of hard delete.
 *   Cascade — Delete/Restore cascade to messages.
 */

import type Dexie from 'dexie';
import type { Transaction } from 'dexie';

import { ErrDuplicateKey, ErrNotFound } from '../errors';
import type { XyncraDatabase } from './index';
import type { Conversation, Message } from './models';

/**
 * ConversationStore provides data access operations for the Conversation model.
 * Mirrors Go ConversationStore (pkg/store/conversation_store.go).
 */
export class ConversationStore {
  constructor(private readonly db: XyncraDatabase) {}

  // ---------------------------------------------------------------------------
  // CRUD
  // ---------------------------------------------------------------------------

  /**
   * Inserts a new conversation record.
   * Throws ErrDuplicateKey if a conversation with the same id already exists.
   */
  async create(conv: Conversation): Promise<void> {
    try {
      await this.db.conversations.add(conv);
    } catch (error) {
      if (this.isConstraintError(error)) {
        throw ErrDuplicateKey;
      }
      throw error;
    }
  }

  /**
   * Retrieves a conversation by its primary key.
   * Returns undefined if not found or soft-deleted.
   */
  async get(id: string): Promise<Conversation | undefined> {
    const conv = await this.db.conversations.get(id);
    // Use != null to match both null and undefined (deleted_at may not be initialized)
    if (!conv || conv.deleted_at != null) {
      return undefined;
    }
    return conv;
  }

  /**
   * Retrieves a conversation including soft-deleted records.
   * Returns undefined if not found at all.
   */
  async getUnscoped(id: string): Promise<Conversation | undefined> {
    return this.db.conversations.get(id);
  }

  /**
   * Returns the 1-on-1 conversation between user1 and user2.
   * Checks both (user1, user2) and (user2, user1) orderings.
   * Returns undefined if no matching conversation exists.
   */
  async getByUsers(
    user1: string,
    user2: string,
  ): Promise<Conversation | undefined> {
    // Try both orderings using the compound unique index.
    let conv = await this.db.conversations
      .where('[user_id1+user_id2]')
      .equals([user1, user2])
      .first();
    if (conv && conv.deleted_at == null) return conv;

    conv = await this.db.conversations
      .where('[user_id1+user_id2]')
      .equals([user2, user1])
      .first();
    if (conv && conv.deleted_at == null) return conv;

    return undefined;
  }

  /**
   * Returns conversations where the given user is either UserID1 or UserID2,
   * ordered by last_message_at descending, with offset/limit pagination.
   * Soft-deleted records are excluded.
   */
  async getByUser(
    userID: string,
    offset: number,
    limit: number,
  ): Promise<Conversation[]> {
    // Database is already scoped by userID+deviceID (xyncra-{userID}-{deviceID}),
    // so we can return all non-deleted conversations without filtering by user_id.
    if (limit <= 0 || limit > 101) limit = 20;
    if (offset < 0) offset = 0;

    // Get all conversations, filter out soft-deleted ones
    // Use == null to match both null and undefined (deleted_at may not be initialized)
    const all = await this.db.conversations
      .filter((conv) => conv.deleted_at == null && !!conv.user_id2)
      .toArray();

    // Sort by last_message_at descending.
    all.sort(
      (a, b) =>
        new Date(b.last_message_at).getTime() -
        new Date(a.last_message_at).getTime(),
    );

    return all.slice(offset, offset + limit);
  }

  /**
   * Saves all fields of the conversation back to the database.
   * Uses put() which handles both insert and update.
   */
  async update(conv: Conversation): Promise<void> {
    await this.db.conversations.put(conv);
  }

  /**
   * Creates the conversation if it does not exist, or saves (overwrites) it if
   * it already exists. Used by the client sync pipeline to apply conversation
   * create events idempotently (D-045).
   *
   * Uses Unscoped() semantics: also finds soft-deleted records, so that
   * restoring a previously deleted conversation correctly transitions it back
   * to active. If a concurrent insert causes a duplicate key error, retries as
   * an update to handle the TOCTOU race.
   */
  async upsert(conv: Conversation): Promise<void> {
    const existing = await this.db.conversations.get(conv.id);
    if (!existing) {
      try {
        await this.db.conversations.add(conv);
        return;
      } catch (error) {
        if (this.isConstraintError(error)) {
          // TOCTOU: another task inserted between our get() and add().
          await this.db.conversations.put(conv);
          return;
        }
        throw error;
      }
    }
    // Record exists (possibly soft-deleted) — overwrite.
    await this.db.conversations.put(conv);
  }

  /**
   * Performs a cascading soft delete: the conversation and all its messages
   * are soft-deleted within a single transaction (D-013).
   * Throws ErrNotFound if the conversation does not exist (or is already deleted).
   */
  async delete(id: string): Promise<void> {
    await this.db.transaction(
      'rw',
      [this.db.conversations, this.db.messages],
      async () => {
        const conv = await this.db.conversations.get(id);
        if (!conv || conv.deleted_at != null) {
          throw ErrNotFound;
        }

        const now = new Date();
        await this.db.conversations.update(id, { deleted_at: now });

        // Cascade soft-delete all messages in this conversation (D-013).
        await this.db.messages
          .where('conversation_id')
          .equals(id)
          .modify((msg) => {
            if (msg.deleted_at == null) {
              msg.deleted_at = now;
            }
          });
      },
    );
  }

  /**
   * Undeletes a soft-deleted conversation and cascades the restore to all its
   * messages within a single transaction (D-015).
   *
   * Idempotent: calling on a non-deleted conversation is a no-op.
   * Throws ErrNotFound if the conversation does not exist at all.
   */
  async restore(id: string): Promise<void> {
    await this.db.transaction(
      'rw',
      [this.db.conversations, this.db.messages],
      async () => {
        const conv = await this.db.conversations.get(id);
        if (!conv) {
          throw ErrNotFound;
        }

        // Restore the conversation if it was soft-deleted.
        if (conv.deleted_at != null) {
          await this.db.conversations.update(id, { deleted_at: null });

          // Cascade restore all messages in this conversation (D-015).
          await this.db.messages
            .where('conversation_id')
            .equals(id)
            .modify((msg) => {
              if (msg.deleted_at != null) {
                msg.deleted_at = null;
              }
            });
        }
      },
    );
  }

  /**
   * Updates the last_message_at and last_processed_message_id fields.
   * Throws ErrNotFound if the conversation does not exist.
   */
  async updateLastMessage(
    convID: string,
    lastMessageAt: Date,
    lastProcessedMessageID: number,
  ): Promise<void> {
    const updated = await this.db.conversations
      .where('id')
      .equals(convID)
      .modify((conv) => {
        conv.last_message_at = lastMessageAt;
        conv.last_processed_message_id = lastProcessedMessageID;
      });
    if (updated === 0) {
      throw ErrNotFound;
    }
  }

  /**
   * Updates the last-read message ID for the specified user.
   * Uses MAX semantics: only advances forward, never backward (C4, D-012).
   *
   * Throws ErrNotFound if the conversation does not exist or the user does
   * not belong to it.
   */
  async updateLastRead(
    convID: string,
    userID: string,
    messageID: number,
  ): Promise<void> {
    await this.db.transaction('rw', this.db.conversations, async () => {
      const conv = await this.db.conversations.get(convID);
      if (!conv) {
        throw ErrNotFound;
      }
      if (conv.user_id1 !== userID && conv.user_id2 !== userID) {
        throw ErrNotFound;
      }

      // MAX semantics: only advance forward.
      const updates: Partial<Conversation> = {};
      if (conv.user_id1 === userID) {
        if (messageID > conv.last_read_message_id1) {
          updates.last_read_message_id1 = messageID;
        }
      }
      if (conv.user_id2 === userID) {
        if (messageID > conv.last_read_message_id2) {
          updates.last_read_message_id2 = messageID;
        }
      }

      if (Object.keys(updates).length > 0) {
        await this.db.conversations.update(convID, updates);
      }
    });
  }

  /**
   * Searches conversations that contain the specified title substring
   * (case-insensitive), ordered by last_message_at descending.
   *
   * Note: Dexie does not support LIKE queries. We perform case-insensitive
   * substring matching in JavaScript after fetching candidates.
   *
   * Note: Database is already scoped by userID+deviceID, so no user_id filtering needed.
   */
  async searchByTitle(
    _userID: string,
    title: string,
    limit: number,
  ): Promise<Conversation[]> {
    if (limit <= 0 || limit > 101) limit = 20;
    if (!title) return [];

    const lowerTitle = title.toLowerCase();

    // Database is single-user scoped; just filter by title.
    const matches = await this.db.conversations
      .filter((conv) => conv.deleted_at == null && conv.title.toLowerCase().includes(lowerTitle))
      .toArray();

    // Sort by last_message_at descending.
    matches.sort(
      (a, b) =>
        new Date(b.last_message_at).getTime() -
        new Date(a.last_message_at).getTime(),
    );

    return matches.slice(0, limit);
  }

  // ---------------------------------------------------------------------------
  // Transaction variants (Tx)
  // ---------------------------------------------------------------------------

  /**
   * Updates last message fields within the given transaction.
   */
  async updateLastMessageTx(
    tx: Transaction,
    convID: string,
    lastMessageAt: Date,
    lastProcessedMessageID: number,
  ): Promise<void> {
    const table = tx.table('conversations') as Dexie.Table<
      Conversation,
      string
    >;
    const updated = await table
      .where('id')
      .equals(convID)
      .modify((conv) => {
        conv.last_message_at = lastMessageAt;
        conv.last_processed_message_id = lastProcessedMessageID;
      });
    if (updated === 0) {
      throw ErrNotFound;
    }
  }

  /**
   * Updates read cursor within the given transaction.
   * Uses MAX semantics: only advances forward (C4, D-012).
   */
  async updateLastReadTx(
    tx: Transaction,
    convID: string,
    userID: string,
    messageID: number,
  ): Promise<void> {
    const table = tx.table('conversations') as Dexie.Table<
      Conversation,
      string
    >;
    const conv = await table.get(convID);
    if (!conv) {
      throw ErrNotFound;
    }
    if (conv.user_id1 !== userID && conv.user_id2 !== userID) {
      throw ErrNotFound;
    }

    const updates: Partial<Conversation> = {};
    if (conv.user_id1 === userID) {
      if (messageID > conv.last_read_message_id1) {
        updates.last_read_message_id1 = messageID;
      }
    }
    if (conv.user_id2 === userID) {
      if (messageID > conv.last_read_message_id2) {
        updates.last_read_message_id2 = messageID;
      }
    }

    if (Object.keys(updates).length > 0) {
      await table.update(convID, updates);
    }
  }

  /**
   * Creates or updates a conversation within the given transaction.
   * Uses Unscoped() semantics (finds soft-deleted records too).
   */
  async upsertTx(tx: Transaction, conv: Conversation): Promise<void> {
    const table = tx.table('conversations') as Dexie.Table<
      Conversation,
      string
    >;
    const existing = await table.get(conv.id);
    if (!existing) {
      try {
        await table.add(conv);
        return;
      } catch (error) {
        if (this.isConstraintError(error)) {
          await table.put(conv);
          return;
        }
        throw error;
      }
    }
    await table.put(conv);
  }

  /**
   * Performs cascading soft delete within the given transaction (D-013).
   */
  async softDeleteTx(tx: Transaction, id: string): Promise<void> {
    const convTable = tx.table('conversations') as Dexie.Table<
      Conversation,
      string
    >;
    const msgTable = tx.table('messages') as Dexie.Table<Message, string>;

    const conv = await convTable.get(id);
    if (!conv || conv.deleted_at != null) {
      throw ErrNotFound;
    }

    const now = new Date();
    await convTable.update(id, { deleted_at: now });

    // Cascade soft-delete messages (D-013).
    await msgTable
      .where('conversation_id')
      .equals(id)
      .modify((msg) => {
        if (msg.deleted_at == null) {
          msg.deleted_at = now;
        }
      });
  }

  /**
   * Performs cascading restore within the given transaction (D-015).
   */
  async restoreTx(tx: Transaction, id: string): Promise<void> {
    const convTable = tx.table('conversations') as Dexie.Table<
      Conversation,
      string
    >;
    const msgTable = tx.table('messages') as Dexie.Table<Message, string>;

    const conv = await convTable.get(id);
    if (!conv) {
      throw ErrNotFound;
    }

    if (conv.deleted_at != null) {
      await convTable.update(id, { deleted_at: null });

      // Cascade restore messages (D-015).
      await msgTable
        .where('conversation_id')
        .equals(id)
        .modify((msg) => {
          if (msg.deleted_at != null) {
            msg.deleted_at = null;
          }
        });
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  /**
   * Checks if an error is a constraint violation (unique index conflict).
   * Dexie throws ConstraintError for unique index violations.
   */
  private isConstraintError(error: unknown): boolean {
    if (error instanceof Error) {
      return error.name === 'ConstraintError';
    }
    return false;
  }
}
